### PR TITLE
Implementation of missing data fields in sheet state engine

### LINKED ISSUE #113: [11] Sheet state engine implement segment fields
Implement the fields regarding the current segment of a `SheetState`. These are currently commented out and have placeholder data assigned to them in the `update()` method.

### PR DESCRIPTION
**Pull request by Extremelyd1**
Fixes #113

#### Deployment instructions
/

#### What have I done
Implemented the missing data fields in the sheet state engine. Includes extra methods in the route singleton class to facilitate the calculation of these fields. The testing of these methods is implemented, but keep in mind that the testing of the `update` method in the `SheetStateEngine` currently contains mocked data and is not elaborate. More in-depth tests will not be implemented in this pull requests, but are addressed in #112.

#### Testing
Automated testing.

### CODE REVIEW

File: src/app/backend/data-representation/route.spec.ts [hunk 1]
```diff
@@ -111,4 +131,165 @@ describe('Route', () => {
...
+    it('should give the correct POI name in another segment', () => {
+        const poiPosition1: POIPosition = new POIPosition(
+            5,
+            new POI(
+                'test1', new ControlPoint(0, 0)
+            )
+        );
+        const segment1: Segment = new Segment(0, 10, [
+            poiPosition1
+        ], []);
```
Comment (TijnBertens) [inline src/app/backend/data-representation/route.spec.ts]: This segment needs multiple POIs, such that you can verify that it picks the last one. I am quite sure this reveals a bug, caused by what I left a comment on in route.ts line 156.

File: src/app/backend/data-representation/route.spec.ts [hunk 2]
```diff
@@ -111,4 +131,165 @@ describe('Route', () => {
...
+
+        const segment2: Segment = new Segment(1, 10, [
+            poiPosition2
+        ], []);
+
+        route.setSegments([segment1, segment2]);
+        route.setSegmentStartPositions([0, 10]);
+
+        expect(route.getLastPOIName(11)).toBe('test1');
```
Comment (TijnBertens) [inline src/app/backend/data-representation/route.spec.ts]: Same as previous comment.

Comment (Extremelyd1) [reply src/app/backend/data-representation/route.spec.ts]: Changed this test case to include another POI in the previous segment to test whether it get the lats one.

File: src/app/backend/data-representation/route.spec.ts (hunk 1 above)
Comment (Extremelyd1) [reply src/app/backend/data-representation/route.spec.ts]: Changed this test case to include another POI in the previous segment to test whether it get the lats one.

File: src/app/backend/data-representation/route.spec.ts [hunk 3]
```diff
@@ -111,4 +131,165 @@ describe('Route', () => {
...
+                new ControlPoint(0, 0)
+            )
+        );
+        const segment1: Segment = new Segment(0, 5, [
+            poiPosition1,
+            poiPosition2,
+            poiPosition3
+        ], []);
+
+        const segment2: Segment = new Segment(1, 12, [
```
[the remark this answers is not in the document]
Comment (Extremelyd1) [reply src/app/backend/data-representation/route.spec.ts]: Changed the test case, now the segments have different POI (with still the same offsets in the segment).

File: src/app/backend/data-representation/route.ts [hunk 4]
```diff
@@ -114,6 +124,45 @@ export class Route {
...
+            poiPositions = currentSegment.getPOIPositions();
+
+            if (poiPositions.length !== 0) {
+                return poiPositions[i].getPOI().getName();
+            }
+        }
+        // Returning undefined, because there is not POI before this position
+        return undefined;
+        // TODO: Might be better to switch to throwing an error. Depends on the input specification.
+        // TODO: (If there always is a POI at the exact start of a route, an error is better)
```
[the remark this answers is not in the document]
Comment (Extremelyd1) [reply src/app/backend/data-representation/route.ts]: Now throws an error if position is negative at the start of the `getLastPOIName` method.

File: src/app/backend/data-representation/route.ts [hunk 5]
```diff
@@ -114,6 +124,45 @@ export class Route {
         throw new IllegalStateError('segment is not in list of segments');
     }
 
+    /**
+     * Returns the last POI from the given position.
+     * @param position The given position.
+     * @return String representation of the name of the last POI before the given position
+     * or undefined as such a POI does not exist.
```
[the remark this answers is not in the document]
Comment (Extremelyd1) [reply src/app/backend/data-representation/route.ts]: Changed documentation to suggestion.

File: src/app/backend/data-representation/route.ts [hunk 6]
```diff
@@ -52,14 +53,23 @@ export class Route {
          */
         for (let i = 1; i < this.segmentStartPositions.length; ++i) {
             if (this.segmentStartPositions[i] >= position) {
-                return this.segments[i - 1];
+                return i - 1;
```
[the remark this answers is not in the document]
Comment (Extremelyd1) [reply src/app/backend/data-representation/route.ts]: Typo has been fixed.

File: src/app/backend/data-representation/route.ts [hunk 7]
```diff
@@ -114,6 +124,45 @@ export class Route {
...
+            }
+        }
+
+        // If the current segment does not contain such a POI, find the last POI in previous segments
+        for (let i = currentSegmentId - 1; i >= 0; i--) {
+            currentSegment = this.segments[i];
+            poiPositions = currentSegment.getPOIPositions();
+
+            if (poiPositions.length !== 0) {
+                return poiPositions[i].getPOI().getName();
```
[the remark this answers is not in the document]
Comment (Extremelyd1) [reply src/app/backend/data-representation/route.ts]: Now indexes the `poiPositions` at last index, thus returning the last POI in an earlier segment. Also added a comment to better explain this.

File: src/app/backend/data-representation/route.ts [hunk 8]
```diff
@@ -114,6 +124,51 @@ export class Route {
...
+     * @throws IllegalArgumentError if the given position is negative.
+     * @return String representation of the name of the last POI before the given position
+     * or undefined if such a POI does not exist.
+     */
+    getLastPOIName(position: number): string {
+        if (position < 0) {
+            throw new IllegalArgumentError('position can not be negative');
+        }
+
+        const currentSegmentId: number = this.getSegmentIndex(position);
```
Comment (marijnvanderhorst) [inline src/app/backend/data-representation/route.ts]: This name is a bit confusing, as it's the index and not the ID. Change to e.g. `currentSegmentIx`

**Pull request merged**